Require to claim revoked local output in its own penalty tx post-anchor (#803)
lightningnetwork/lightning-rfc: 05-onchain.md

  using the revocation private key.
  - SHOULD extract the payment preimage from the transaction input witness, if
  it's not already known.
  - if `option_anchor_outputs` applies:

>     - MUST *resolve* the _local node's main output_ in its own penalty transaction

Reviewer: Perhaps this should be "should" instead? As there're other ways to defend against this, for example: sweep every output in its own transaction.

Author: Reworded, it's used now a `MAY use a single transaction to *resolve* all the outputs` which by the negative allow to sweep every output in its own transaction as _first_ broadcast to mitigate. I kept the MUST if the confirmation isn't reached near to CSV expiration as suggested by Johan. Thus you're free to aggregate as you like until some point to save on fees but after it's a matter of security and should you care to confirm.

  using the revocation private key.
  - SHOULD extract the payment preimage from the transaction input witness, if
  it's not already known.
  - if `option_anchor_outputs` applies:

>   - otherwise:

Reviewer 2: Looks like you're missing something there...?

    - if confirmation doesn't happen before reaching `security_delay` blocks from
  expiry:

>       - MUST *resolve* the _local node's `to_local` output_ in its own, separate penalty transaction   A previous penalty transaction claiming multiple revoked outputs at once may be blocked from

Reviewer 2: nit:

```suggestion
      - MUST *resolve* the _local node's `to_local` output_ in its own, separate penalty transaction. A previous penalty transaction claiming multiple revoked outputs at once may be blocked from
```

>       - MUST *resolve* the _local node's main output_ in its own penalty transaction
>   as if a single resolving-transaction has been previously used its propagation might be
>   under pinning.

Reviewer 2: nit:

```suggestion
      - MUST *resolve* the _local node's main output_ in its own, separate penalty transaction.
      A previous penalty transaction claiming multiple revoked outputs at once may be blocked
      from confirming because of a transaction pinning attack.
```

  using the revocation private key.
  - SHOULD extract the payment preimage from the transaction input witness, if
  it's not already known.
  - if `option_anchor_outputs` applies:
    - MAY use a single transaction to *resolve* all the outputs.
    - if confirmation doesn't happen before reaching `security_delay` blocks from
  expiry:

>       - MUST *resolve* revoked outputs in their own, separate penalty transactions. A previous

Reviewer 3: Feels like `SHOULD` is more correct here, since the sweeping node is aware of the risk and should be allowed to do as it pleases.

Reviewer 2: I agree, @ariard I think this is the last pending point before we merge, what do you think?

Reviewer 4: We've got the ACKs on this, should we wait til this is addressed? paging @ariard

Author: Sounds good to me, it's up to the implementer to gauge risk-vs-fee-saving.

standard size limit because of the 483 HTLC-per-party limit (see
[BOLT #2](02-peer-protocol.md#the-open_channel-message)).

Note: if `option_anchor_outputs` applies, the cheating node can pin spends of its
HTLC-timeout/HTLC-success outputs thanks to SIGHASH_SINGLE malleability.
Using a single penalty transaction for all revoked outputs is thus unsafe as it

> could be blocked to propagate long enough for the `_local node's main output_ 's

Reviewer: IMO it's worth nothing more of the subtly here. If they're pinning, then in order to even get into the mempool, they're likely paying a relatively high fee, which means that if they _do_ confirm, then there's the CSV period on the second level. Also the defender can mix things up themselves, by only sweeping _one_ of the HTLC outputs, which would then invalidate the entire "tree" (possibly needing to pay high enough fees?) of the attacker.

Author: As usual they might pay a high fee but low-feerate, which isn't a real deterrence. Updated with the reference to CSV period. You can sweep one of the HTLC output thus invalidating the batched pinning of the attacker, but the attacker can pinning each commitment HTLC output at first broadcast as you have full-malleability for each of them. Generally I don't think we should successively "react" compared on guessed attacker moves as our transaction broadcast algorithm is public but not attacker one.

standard size limit because of the 483 HTLC-per-party limit (see
[BOLT #2](02-peer-protocol.md#the-open_channel-message)).

Note: if `option_anchor_outputs` applies, the cheating node can pin spends of its
HTLC-timeout/HTLC-success outputs thanks to SIGHASH_SINGLE malleability.
Using a single penalty transaction for all revoked outputs is thus unsafe as it

> could be blocked to propagate long enough for the `_local node's main output_ 's

Reviewer 2: ```suggestion
could be blocked to propagate long enough for the `_local node's main output_`s
```
